### ISSUE TITLE
Trigger new request using moockoon

### ISSUE THREAD

**Issue by eliseuborges** [mockoon/mockoon#709]
Hi @255kb, 

Nowadays is very common async communication between micro services like the scenario bellow:

**AppA** requests **AppB** and **AppB** responds with http 202.
After processing (X seconds), **AppB** responds the process result, on a new request, to **AppA**.

Using moockon to reduce integration test code, I think that could be nicer if it provides some way to trigger or execute that new request to specific host:port and after X seconds, simulating above case.

What do you think?

**Comment by eliseuborges** [comment]
OK Guillaume! Tks!

Just to complement... I found this page, that represents how I imagine this feature: https://apisimulator.io/docs/latest/standalone-api-simulator/callback.html

**Comment by shafihuzaib** [comment]
This would be a necessary feature in the coming future!

**Comment by isuru89** [comment]
Would love to see this feature soon, because we need to simulate webhook scenarios.

**Comment by isuru89** [comment]
I have a prototype working solution in my fork. Which callbacks are defined as below.

  * Callbacks are scoped under a Response to allow difference responses to trigger different callbacks
  * Callbacks are completely optional and also can have as many as callbacks for a response.
  * Callback can have a description.
  * Callbacks must have HTTP verb, the target url, and optional headers / body.
  * The body can be a inline, based on file, or based on a data bucket, very similar to the existing response body section.
  * Callbacks under a response can easily be added, cloned or deleted.
  * Callbacks will be executed asynchronously, just before the response is returned
  * Optionally, callbacks can have a latency, so that those will be called after the specified delay (in ms)
  * Order of the execution of callbacks is dependent on specified order unless having a latency defined in each.

**Current limitations:**
  1. Callbacks and their call status will not be available in Logs, due to their asynchronous nature.
  2. Synchronous callbacks are not supported yet.
  3. Callbacks cannot cloned across different responses

Love to hear some feedback on this.

**WIP UI:**

![mockoon-callbacks](https://user-images.githubusercontent.com/1499404/236190582-e82155e7-8c6a-4187-a7be-f1d36e81844c.png)

**Comment by 255kb** [comment]
@isuru89 Thanks for exploring this. If you are willing to work on this feature (including tests, etc) let me know and I will assign it to you. It's initially something I really didn't want to add in Mockoon (I wanted the app to focus on responses solely), but it seems to have gathered some interest and I can see the use case. 

It's quite a big feature and if we want to add this we have to do it right and make sure the usability is top notch. 

Concerning the current prototype, I would say it's too much in a place where usually we only have response information. 
I would create a new tab at the environment level (next to "Logs" or "Proxy") where you can create the callbacks, with an interface similar to the Data buckets: list of callbacks on the left, callbacks params on the right. 

Then inside the route response, we could add this "Callbacks" tab, with only a list of callbacks, with an interface a bit similar to the headers or rules: 
- Add callback button 
- Dropdown to choose the callback
- Trash can to remove the callback

Nice to have: 
- Reorder the callbacks

Questions: 
- latency should maybe be at the response level?

This would allow for callbacks reusability and to declutter the response interface. 

If you want, we can discuss this more in details on the [Discord server](https://discord.com/invite/JmTBk59MTb).

**Comment by isuru89** [comment]
@255kb Yes, please. I would love to contribute to this feature and assign it to me. 
What you proposed here is better than mine. While I was implementing everything under new Callback tab, I realized it is not elegant (although it works). I was also thinking about a similar approach like data buckets. And currently experimenting with it.

> Nice to have:
>   * Reorder the callbacks
This is possible.

> Questions:
>   * latency should maybe be at the response level?

Yes, because, I have seen that sometimes the latency may differ for different service end points to simulate different scenarios.

**Comment by 255kb** [comment]
@isuru89 I assigned this issue to you and put it on the [roadmap](https://github.com/orgs/mockoon/projects/9). 
Let me know if you need anything here on Discord.

**Comment by viralpraxis** [comment]
Hi @isuru89! I wonder if there is some progress on this feature?

Also, shouldn't callbacks bodies support templating? I believe It would be extremely useful to be able to generate bodies using original request's parameters. Say, if I made a request with `price: 42` parameter I expect the callback to include `total_price: 42` in it's body.

**Comment by eliseuborges** [comment]
Hey @isuru89 do you still working on this feature?

**Comment by isuru89** [comment]
This is my current working branch. https://github.com/isuru89/mockoon/tree/feature/709-response-callbacks
I am almost done except for writing few more tests for response callbacks scenarios.

@viralpraxis Yes, templating will be supported in response bodies.

**Comment by isuru89** [comment]
@255kb  
I have followed your guide as mentioned above. And now I have a completed version of response callbacks. This implementation is very similar to the databuckets where you can define all your callbacks under an environment and then those can be attached to each response rule.

I will raise a PR with details. However, the templating specs are failing in the forked repository. @255kb I need more information why its failing and I don't think it is because these callback changes.

This is how it looks like:
1. A user can define callbacks for an environment
![env-callbacks](https://github.com/mockoon/mockoon/assets/1499404/cf0075a7-5c47-4f89-8ceb-747b78cb7eaa)

2. A user can then attach those callbacks under a response rule.
![response-callbacks](https://github.com/mockoon/mockoon/assets/1499404/2606ad0d-8639-4426-8030-cee7e5a7e8d5)

**Comment by 255kb** [comment]
@isuru89 This looks awesome. It is also a huge amount of work for which I am grateful :)
From what I see in the [last action](https://github.com/isuru89/mockoon/actions/runs/6649949672/job/18069267854), error is coming from [this line](https://github.com/mockoon/mockoon/compare/main...isuru89:mockoon:feature/709-response-callbacks#diff-0795ec89d79ac105c1680f17b98ad74d7bc7ca97fa9d7e6b1916e634accf4f07R1450). Not sure what the problem is exactly, but I see that your [environment test file](https://github.com/mockoon/mockoon/compare/main...isuru89:mockoon:feature/709-response-callbacks#diff-c7b585c54121049245cb4b0493ee633b1ea1c15f986687a919a060ed0aa38eeaR3) has callbacks but is still in migration 27. 

Some remarks from the screenshots (but it's better to continue discussing these in a PR I think): 
- I would prefer another icon to navigate inside the app (I don't have an idea yet): 
  ![image](https://github.com/mockoon/mockoon/assets/7489814/684d11bf-ccf8-47a7-8f16-1876d7e75b5e)
  I think this icon is rather symbolizing a navigation outside, and it is its current use. 
- Why not reusing the same HTTP methods menu rather than a new one?
- After removing support for Node16 in v5.0.0, we got rid of the axios dependency everywhere in profit of `fetch` which is now available by default. I would prefer we avoid reintroducing this dependency. 
- you added [new filter actions and properties](https://github.com/mockoon/mockoon/compare/main...isuru89:mockoon:feature/709-response-callbacks#diff-b80c788f3b335132f26c2bca723cd2cf6086a12120cdf861309ad2edb99f190bR61-R63) to the store. But this was the role of the `store.filters` object which you should use. 

I suggest you open a PR, even if everything is not ready yet, to start the review together.

**Comment by simonhitachi** [comment]
See my discussion around using https in callbacks: https://github.com/mockoon/mockoon/discussions/1272
